### PR TITLE
feat(appmesh): change Route's spec to a union-like class

### PR DESCRIPTION
**Pull request by dfezzie**
Implements the same pattern `GatewayRoutes` use for defining protocol specific specs.

Adds GRPC and HTTP2 support to Routes which will resolve https://github.com/aws/aws-cdk/issues/9755

BREAKING CHANGE: changes Route's spec to a union-like class. RouteSpec is now defined using protocol variant static methods

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+   * @default - no http2 spec
+   */
+  readonly http2RouteSpec?: CfnRoute.HttpRouteProperty;
+
+  /**
+   * The spec for a grpc route
+   *
+   * @default - no grpc spec
+   */
+  readonly grpcRouteSpec?: CfnRoute.GrpcRouteProperty;
```
Comment (dfezzie) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:93]: Could easily implement this in this PR, but I figure it may be better in a separate PR

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+   *
+   * @default 1
+   */
+  readonly weight?: number;
+}
+
+/**
+ * The criterion for determining a request match for this GatewayRoute
+ */
+export interface HttpRouteMatch {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:26]: This is exactly the same as the existing `HttpRouteMatch` interface. Perhaps we should have only one interface for this?

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 3]
```diff
@@ -195,39 +195,63 @@ The listeners property can be left blank and added later with the `node.addListe
...
+```
+
+The _RouteSpec_ class provides an easy interface for defining new protocol specific route specs.
+The `tcp()`, `http()` and `http2()` methods provide the spec necessary to define a protocol specific spec.
+
+For HTTP based routes, the match field can be used to match on a route prefix.
+By default, an HTTP based route will match on `/`. All matches must start with a leading `/`.
+
+```typescript
+router.addRoute('route-http', {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/README.md:244]: This is pretty much identical to the example on line 201... maybe show something else in the ReadMe, like a GRPC RouteSpec?

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  readonly weightedTargets: WeightedTarget[];
+
+  /**
+   * The priority for the route. Routes are matched based on the specified value, where 0 is the highest priority.
+   *
+   * @default - none
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:55]: What does it mean for a route to not have a priority? Does this mean all routes without this set have the same, lowest, priority? Or something else? Feels like this needs more of an explanation here.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:93]: Sorry, what is "this" here? 🙂

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  abstract readonly weightedTargets: WeightedTarget[];
+
+  /**
+   * The priority for the route. Routes are matched based on the specified value, where 0 is the highest priority
+   */
+  abstract readonly priority?: number;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:144]: Why are we exposing these - I don't see them used in `Route`...? I think we only need `bind()` in this class, no?

We should strive to minimize the public API surface as much as we can.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  /**
+   * Called when the GatewayRouteSpec type is initialized. Can be used to enforce
+   * mutual exclusivity with future properties
+   */
+  public abstract bind(scope: cdk.Construct): RouteSpecConfig;
+
+  /**
+  * Utility method to add weighted route targets to an existing route
+  */
+  protected renderWeightedTargets() {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:155]: Can you just make this a function in this file that takes the array of `WeightedTarget`s as an argument?

File: packages/@aws-cdk/aws-appmesh/test/test.route.ts [hunk 7]
```diff
@@ -1,9 +1,200 @@
...
+              ],
+            },
+          },
+        },
+        RouteName: 'test-tcp-route',
+      }));
+
+      test.done();
+    },
+    'should have expected features'(test: Test) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/test.route.ts:133]: There should be an empty line between tests on the same level:

```suggestion
    },

    'should have expected features'(test: Test) {
```

Same comment everywhere in the tests.

**Review by skinny85 — DISMISSED** [review]
Looks good! Couple of notes.

You're also missing a BREAKING CHANGE in your PR description.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 1 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:93]: GRPC Route Specs

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:93]: I'm fine with adding them in this PR 🙂

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 5 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:144]: So, this field makes a bit more sense once we implement header based routing for HTTP.

Essentially it is used to disambiguate when a request matches multiple routes. I'll remove for this PR and we can add it back later!

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 6 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:155]: Any reason why? In this comment you mentioned the `Listener` class should be responsible for rendering itself: https://github.com/aws/aws-cdk/pull/10879#discussion_r50510394

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 4 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:55]: going to implement this later, removing for now

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 2 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:26]: This is actually only the case because we have not implemented the full `RouteMatch` for routes yet:

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-route-httproutematch.html

They do differ on the underlying CFN resource between gateway routes and routes. This could change in the future, but right now it is the case. I don't see a way this could be handled with a single `HttpRouteMatch` interface.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:26]: OK. I guess we changed something in the other PR to not use the name `HttpRouteMatch` anymore...?

We can't have 2 interfaces exported with the same name from one module.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+/**
+ * Properties specific for a GRPC Based Routes
+ */
+export interface GrpcRouteSpecOptions {
+  /**
+   * The criterion for determining a request match for this Route
+   *
+   * @default - no default
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:80]: This property is not optional...?

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 6 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:170]: Having a file-private function that the different `RouteSpec` classes use does not change that (it's an implementation detail).

What I don't want is an additional `protected` (which is effectively the same as `public`) method on this class. We should strive to keep the API of these classes to a minimum.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+   * Creates a GRPC Based RouteSpec
+   */
+  public static grpc(options: GrpcRouteSpecOptions): RouteSpec {
+    return new GrpcRouteSpec(options);
+  }
+
+  /**
+   * List of targets that traffic is routed to when a request matches the route
+   */
+  abstract readonly weightedTargets: WeightedTarget[];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:159]: Let's get rid of this property. The public API of this class should be the minimum we need.

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json [hunk 10]
```diff
@@ -599,6 +599,49 @@
...
+          }
+        },
+        "VirtualRouterName": {
+          "Fn::GetAtt": [
+            "meshrouter81B8087E",
+            "VirtualRouterName"
+          ]
+        }
+      }
+    },
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json:644]: Hmm, why changes to this file? I understand the API changed, but after updating the code to use it, the resulting template should be the same as before...

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts [hunk 11]
```diff
@@ -99,22 +103,42 @@ const node3 = mesh.addVirtualNode('node3', {
...
+      {
+        virtualNode: node3,
+        weight: 20,
+      },
+    ],
+    match: {
+      serviceName: virtualService.virtualServiceName,
     },
-  ],
+  }),
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:141]: Let's keep this file as close to the original as possible (only make changes to make the code compile with the new APIs).

**Review by skinny85 — DISMISSED** [review]
Looks good! Some minor comments.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 2 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:26]: The other one is called `HttpGatewayRouteMatch` :)

File: packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json (hunk 10 above)
Comment (dfezzie) [reply packages/@aws-cdk/aws-appmesh/test/integ.mesh.expected.json:644]: Added the other route types to the integ file. Removing them

**Review by skinny85 — APPROVED** [review]
Looks fantastic, thanks @dfezzie !
